librad2/src/git/mod.rs

```diff
@@ -0,0 +1,287 @@
+use std::fmt;
+use std::io;
+use std::str::FromStr;
+
+use git2;
+use olpc_cjson::CanonicalFormatter;
+use serde::Serialize;
+use serde_json;
+
+use crate::keys;
+use crate::keys::device;
+use crate::keys::pgp;
+use crate::meta::project as meta;
+use crate::paths::Paths;
+use crate::peer::PeerId;
+use crate::project::Project;
+
+const PROJECT_METADATA_REF: &str = "refs/heads/rad/project";
+const PROJECT_METADATA_BRANCH: &str = "rad/project";
+const PROJECT_METADATA_FILE: &str = "project.json";
+const RAD_REMOTE_NAME: &str = "rad";
+
+#[derive(Debug, Fail)]
+pub enum Error {
+    #[fail(display = "Invalid key")]
+    InvalidKey,
+
+    #[fail(display = "Project already exists")]
+    ProjectExists,
+
+    #[fail(display = "Project not found")]
+    NoSuchProject,
+
+    #[fail(display = "Unable to determine project name from repository directory")]
+    NeedProjectDir,
+
+    #[fail(display = "{}", 0)]
+    InvalidProjectId(projectid::ParseError),
+
+    #[fail(display = "{}", 0)]
+    Libgit(git2::Error),
+
+    #[fail(display = "{}", 0)]
+    Io(io::Error),
+
+    #[fail(display = "{}", 0)]
+    Serde(serde_json::error::Error),
+
+    #[fail(display = "{}", 0)]
+    Pgp(keys::pgp::Error),
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Self {
+        Error::Io(err)
+    }
+}
+
+impl From<git2::Error> for Error {
+    fn from(err: git2::Error) -> Self {
+        Error::Libgit(err)
+    }
+}
+
+impl From<serde_json::error::Error> for Error {
+    fn from(err: serde_json::error::Error) -> Self {
+        Error::Serde(err)
+    }
+}
+
+impl From<keys::pgp::Error> for Error {
+    fn from(err: keys::pgp::Error) -> Self {
+        Error::Pgp(err)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ProjectId(git2::Oid);
+
+pub mod projectid {
```

Comment:
What's the reason for this `mod`?

Reply:
Namespacing :) I could’ve also written `ProjectIdParseError`, but you know those camel bumps

Reply:
Ah ok, that makes sense!